tools/print_tock_memory_usage.py

```diff
@@ -0,0 +1,422 @@
+                    kernel_functions.append((name, addr, size, 0))
+
+def print_section_information():
+    """Print out the ELF's section information (RAM and Flash use)."""
+    text_size = sections["text"]
+    stack_size = sections["stack"]
+    relocate_size = sections["relocate"]
+    sram_size = sections["sram"]
+    app_size = 0
+    if "app_memory" in sections:  # H1B-style linker file, static app section
+        app_size = sections["app_memory"]
+    else: # Mainline Tock-style linker file, using APP_MEMORY
+        for (name, addr, size, tsize) in kernel_uninitialized:
+            if name.find("APP_MEMORY") >= 0:
+                app_size = size
+
+    flash_size = text_size + relocate_size
+    ram_size = stack_size + sram_size + relocate_size
+
+    print("Kernel occupies " + str(flash_size) + " bytes of flash")
+    print("  " + "{:>6}".format(text_size) + "\tcode and constant strings")
+    print("  " + "{:>6}".format(relocate_size) + "\tvariable initializers")
+    print("Kernel occupies " + str(ram_size) + " bytes of RAM")
+    print("  " + "{:>6}".format(stack_size) + "\tstack")
+    print("  " + "{:>6}".format(sram_size) + "\tuninitialized variables")
+    print("  " + "{:>6}".format(relocate_size) + "\tinitialized variables")
+    print("  " + "{:>6}".format(sram_size + relocate_size) + "\tvariables total")
+    print("Applications allocated " + str(app_size) + " bytes of RAM")
+
+# Take a list of 'symbols' and group them into in 'groups' as aggregates
+# for condensing. Names are '::' delimited hierarchies. The aggregate
+# sizes are determined by the global symbol depth, which indicates how
+# many levels of the naming heirarchy to display. A depth of 0 means
+# group all symbols together into one category; a depth of 1 means
+# aggregate symbols into top level categories (e.g, 'h1b::*'). A depth
+# of 100 means aggregate symbols only if they have the same first 100
+# name levels, so effectively print every symbol individually.
+#
+# The 'waste' and 'section' parameters are used to specify whether detected
+# waste should be printed and the name of the section for waste information.
+def group_symbols(groups, symbols, waste, section):
+    """Take a list of symbols and group them into 'groups' for reporting
+       aggregate flash/RAM use."""
+    global symbol_depth
+    expected_addr = 0
+    waste_sum = 0
+    prev_symbol = ""
+    for (symbol, addr, size, _) in symbols:
+        if size == 0:
+            continue
+        # If we find a gap between symbol+size and the next symbol, we might
+        # have waste. But this is only true if it's not the first symbol and
+        # this is actually a variable and just just a symbol (e.g., _estart)
+        if addr != expected_addr and expected_addr != 0 and size != 0 and (waste or verbose):
+            print("  ! " + str(addr - expected_addr) + " bytes wasted after " + prev_symbol)
+        waste_sum = waste_sum + (addr - expected_addr)
```

Comment:
Just tried this out, I think this should be under the 'if' as well, otherwise it throws off the output:

```
Total of 268439552 bytes wasted in RAM
  ! 3 bytes wasted after components..lldb..LowLevelDebugComponent
  ! 344 bytes wasted after opentitan::reset_handler::BUF
Total of 268443655 bytes wasted in Flash+RAM
```

Reply:
Nice catch!

Reply:
Fixed.